src/ros_tcp_endpoint/tcp_sender.py

```diff
@@ -19,108 +19,103 @@
-        message = UnityHandshake._request_class().deserialize(data)
-        self.unity_port = message.port
-        if not self.ip_is_overridden:
-            # hello Unity, we'll talk to you from now on
-            if message.ip == '':
-                # if the message doesn't specify an IP, just talk back to the incoming IP address
-                self.unity_ip = incoming_ip
-            else:
-                # otherwise Unity has set an IP override, so use that
-                self.unity_ip = message.ip
-        print("ROS-Unity Handshake received, will connect to {}:{}".format(self.unity_ip, self.unity_port))
-        return UnityHandshakeResponse(self.unity_ip)
+        self.sender_id = 1
+        
+        # variables needed for matching up unity service requests with responses
+        self.next_srv_id = 1001
+        self.srv_lock = threading.Lock()
+        self.services_waiting = {}
 
     def send_unity_error(self, error):
         self.send_unity_message("__error", RosUnityError(error))
 
     def send_unity_message(self, topic, message):
-        if self.unity_ip == '':
-            print("Can't send a message, no defined unity IP!".format(topic, message))
-            return
-
         serialized_message = ClientThread.serialize_message(topic, message)
         self.queue.put(serialized_message)
 
     def send_unity_service(self, topic, service_class, request):
-        if self.unity_ip == '':
-            print("Can't send a message, no defined unity IP!".format(topic, request))
-            return
-
-        serialized_message = ClientThread.serialize_message(topic, request)
+        request_bytes = BytesIO()
+        request.serialize(request_bytes)
+        result = [None]
+        condition = threading.Condition()
         
-        try:
-            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            s.settimeout(self.timeout_on_connect)
-            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            s.connect((self.unity_ip, self.unity_port))
-            s.settimeout(self.timeout_on_send)
-            s.sendall(serialized_message)
-
-            destination, data = ClientThread.read_message(s)
+        with self.srv_lock:
+            srv_id = self.next_srv_id
+            self.next_srv_id+=1
+            self.services_waiting[srv_id] = (condition, result)
+        
+        payload = request_bytes.getvalue()
+        srv_message = RosUnitySrvMessage(srv_id, True, topic, payload)
+        serialized_message = ClientThread.serialize_message('__srv', srv_message)
+        self.queue.put(serialized_message)
 
-            response = service_class._response_class().deserialize(data)
+        # rospy starts a new thread for each service request,
+        # so it won't break anything if we suspend it while waiting for the response
+        with condition:
+            condition.wait()
+        
+        response = service_class._response_class().deserialize(result[0])
```

Comment:
lol ah ok I was wondering about the result in a List... this should work but I feel like you could enforce the coupling of a `condition` to a specific `result` better by just defining a class which holds one of each and storing references to those class instances in your dictionary.  Having a bunch of `result[0]` around the code seems like the type of thing that we'd come back to six months from now and be totally confounded by.